e2e test for dynamic provisioning.

jsafrane opened the pull request:
The test works only on GCE, AWS and OpenStack. It creates a 2GiB volume, creates a pod that writes to it, creates another pod that read from it and then it deletes the volume.

It runs on current master **without** PersistentVolumeLabel  admission controller (#21041).

~~It produces flakes #20213 and #19868~~ - fixes have been just merged

jsafrane commented:
Fixes #20564

jsafrane commented:
Especially note there is one magic sleep with a long comment why it's there, I hope it's enough.

jsafrane commented:
@kubernetes/rh-storage

jsafrane commented:
> GCE e2e build/test failed 

That's buggy PersistentVolumeLabel admission controller (#21041).

ihmccreery commented:
Reassigning to `team/cluster`.

jsafrane commented:
Rebased

jsafrane committed:
e2e test for dynamic provisioning.

The test works only on GCE, AWS and OpenStack. It creates and deletes a 2GiB
volume.

saad-ali commented on test/e2e/volume_provisioning.go:
```diff
@@ -0,0 +1,197 @@
+			claim, err := c.PersistentVolumeClaims(ns).Create(claim)
+			Expect(err).NotTo(HaveOccurred())
+
+			err = waitForPersistentVolumeClaimPhase(api.ClaimBound, c, ns, claim.Name, 1*time.Second, 30*time.Second)
```
Is 30 seconds enough time for all cloud providers to create a volume. I know attach/detach can sometimes take a few minutes to complete.

jsafrane replied:
In my experience, creation was always almost instant operation. Attachment was much slower. Still, we may hit some rate limiter in AWS, I'll increase it to 5 minutes.

saad-ali commented on test/e2e/volume_provisioning.go:
```diff
@@ -0,0 +1,197 @@
+		expectNoError(c.Pods(ns).Delete(pod.Name, nil))
+	}()
+	expectNoError(err, "Failed to create pod: %v", err)
+	expectNoError(waitForPodSuccessInNamespace(c, pod.Name, pod.Spec.Containers[0].Name, pod.Namespace))
```
Might wanna use `waitForPodRunningInNamespaceSlow` here. Attach/detach can sometimes take several minutes to complete. See `test/e2e/pd.go` tests, for example.

jsafrane replied:
15 minutes? Ok, I trust your judgment.

saad-ali replied:
Yep: https://github.com/kubernetes/kubernetes/issues/15382#issuecomment-153268655
We need to follow up with the GCE folks on that.

saad-ali commented on test/e2e/volume_provisioning.go:
```diff
@@ -0,0 +1,197 @@
+			By("checking the created volume is readable and retains data")
+			runInPodWithVolume(c, ns, claim.Name, "grep 'hello world' /mnt/test/data")
+
+			// Ugly hack: if we delete the AWS/GCE/OpenStack volume here, it will
```
Not for this PR. But this reminds me that we need to think through the logic that triggers deletion of auto-provisioned volumes. Opened up https://github.com/kubernetes/kubernetes/issues/22975

saad-ali commented on test/e2e/volume_provisioning.go:
```diff
@@ -0,0 +1,197 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
```
Update copyright year on new files.

saad-ali commented:
LGTM mod nits

jsafrane commented:
Updated:
* added `claimProvisionTimeout` to test/e2e/util.go
* used `poll` for polling frequency
* used `slowPodStartTimeout` (=15 minutes) for pod startup
* and reworked `waitForPodSuccessInNamespace` - new `waitForPodSuccessInNamespaceSlow` created out of it.

(it looks like huge list, but it's bunch of one-liners)

saad-ali commented:
Thanks @jsafrane 
LGTM

saad-ali commented:
Now that this is merged, let's let it bake and see how stable it is.
If it is stable enough, we can cherry-pick it to the v1.2 branch.

zmerlynn commented:
@eparis: Here's another where @k8s-cherrypick-bot didn't remove the `cherrypick-candidate`. I'm going to start manually removing these.
